File: cpp/3_6.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>

using namespace std;

// Функция для заполнения вектора случайными целыми числами от 1 до 100
void fillVectorWithRandom(vector<int>& vec) {
    random_device rd;
    ranlux48_base gen(rd());  // Создаем генератор случайных чисел ranlux48_base
    uniform_int_distribution<int> dis(1, 100);  // Определяем распределение от 1 до 100

    for (int& num : vec) {
        num = dis(gen);  // Заполняем вектор случайными числами
    }
}

// Функция для подсчета количества вхождений элемента в векторе
int countOccurrences(const vector<int>& container, int element) {
    int count = 0;
    for (int value : container) {
        if (value == element) {
            count++;
        }
    }
    return count;
}

// Функция для вычисления среднего значения элементов вектора
double mean(const vector<int>& container) {
    double n = 0;
    for (int i : container){
        n += i;
    }
    return n/container.size();
}

// Функция для вычисления дисперсии элементов вектора
double variance(const vector<int>& container) {
    double m = mean(container);  // Вычисляем среднее значение элементов вектора
    double sum_sq_diff = 0.0;
    for (int value : container) {
        sum_sq_diff += (value - m) * (value - m);  // Суммируем квадраты отклонений от среднего
    }
    return sum_sq_diff / (container.size() - 1);  // Возвращаем дисперсию
}

// Функция для вычисления функции нормального распределения (Cumulative Distribution Function - CDF)
double normalCDF(double x) {
    return (1.0 / 2.0) * (1.0 + erf(x / sqrt(2.0)));  // Возвращаем значение CDF
}

// Функция для проведения теста хи-квадрат на нормальность выборки
double chiSquareTest(const vector<int>& container) {
    int size = container.size();  // Размер выборки
    double chi2 = 0;
    double mean_val = mean(container);  // Среднее значение выборки
    double std_dev = sqrt(variance(container));  // Стандартное отклонение выборки

    for (int i = 1; i <= 100; i++) {
        // Вычисляем ожидаемую частоту для интервала [i-1, i)
        double expected_freq = size * (normalCDF((i - mean_val) / std_dev) - normalCDF((i - 1 - mean_val) / std_dev));
        // Вычисляем наблюдаемую частоту (количество вхождений числа i в выборку)
        double observed_freq = countOccurrences(container, i);
        // Вычисляем величину хи-квадрат
        chi2 += pow(observed_freq - expected_freq, 2) / expected_freq;
    }

    return chi2;  // Возвращаем значение статистики хи-квадрат
}

// Функция для вывода результатов теста хи-квадрат
void chiPrint(const vector<int>& vec, double& criticalValue) {
    double chiSquareValue = chiSquareTest(vec);  // Вычисляем статистику хи-квадрат
    double meanExpected = mean(vec);  // Вычисляем ожидаемое среднее значение
    double meanObserved = mean(vec);  // Вычисляем наблюдаемое среднее значение

    // Выводим результаты теста хи-квадрат и сравниваем с критическим значением
    cout << "Chi-Square P-value: " << chiSquareValue << endl;
    if (chiSquareValue <= criticalValue) {
        cout << "Hypothesis of normal distribution is not rejected." << endl;
    } else {
        cout << "Hypothesis of normal distribution is rejected." << endl;
    }
    cout << "Expected mean: " << meanExpected << endl;
    cout << "Observed mean: " << meanObserved << endl;
    cout << endl;
}

int main() {
    // Создаем векторы различных размеров и заполняем их случайными числами
    vector<int> vector50(50);
    vector<int> vector100(100);
    vector<int> vector1000(1000);
    fillVectorWithRandom(vector50);
    fillVectorWithRandom(vector100);
    fillVectorWithRandom(vector1000);

    double criticalValue = 124.342;  // Критическое значение для теста хи-квадрат
    // Проводим тест хи-квадрат и выводим результаты для каждого вектора
    chiPrint(vector50, criticalValue);
    chiPrint(vector100, criticalValue);
    chiPrint(vector1000, criticalValue);

    return 0;
}

File: go/3_6.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

func fillVectorWithRandom(vec []int) {
	for i := range vec {
		vec[i] = rand.Intn(100) + 1
	}
}

func countOccurrences(container []int, element int) int {
	count := 0
	for _, value := range container {
		if value == element {
			count++
		}
	}
	return count
}

func mean(container []int) float64 {
	var n float64
	for _, i := range container {
		n += float64(i)
	}
	return n / float64(len(container))
}

func variance(container []int) float64 {
	m := mean(container)
	var sumSqDiff float64
	for _, value := range container {
		sumSqDiff += (float64(value) - m) * (float64(value) - m)
	}
	return sumSqDiff / float64(len(container)-1)
}

func normalCDF(x float64) float64 {
	return (1.0 / 2.0) * (1.0 + math.Erf(x/math.Sqrt2))
}

func chiSquareTest(container []int) float64 {
	size := len(container)
	var chi2 float64
	meanVal := mean(container)
	stdDev := math.Sqrt(variance(container))

	for i := 1; i <= 100; i++ {
		expectedFreq := float64(size) * (normalCDF(float64(i-int(meanVal))/stdDev) - normalCDF(float64(i-1-int(meanVal))/stdDev))
		observedFreq := float64(countOccurrences(container, i))
		chi2 += math.Pow(observedFreq-expectedFreq, 2) / expectedFreq
	}

	return chi2
}

func chiPrint(vec []int, criticalValue float64) {
	chiSquareValue := chiSquareTest(vec)
	meanExpected := mean(vec)
	meanObserved := mean(vec)

	fmt.Println("Chi-Square P-value:", chiSquareValue)
	if chiSquareValue <= criticalValue {
		fmt.Println("Hypothesis of normal distribution is not rejected.")
	} else {
		fmt.Println("Hypothesis of normal distribution is rejected.")
	}
	fmt.Println("Expected mean:", meanExpected)
	fmt.Println("Observed mean:", meanObserved)
	fmt.Println()
}

func main() {
	vector50 := make([]int, 50)
	vector100 := make([]int, 100)
	vector1000 := make([]int, 1000)
	fillVectorWithRandom(vector50)
	fillVectorWithRandom(vector100)
	fillVectorWithRandom(vector1000)
	criticalValue := 124.342 // Critical chi-square value for 99 degrees of freedom and 0.05 significance level
	chiPrint(vector50, criticalValue)
	chiPrint(vector100, criticalValue)
	chiPrint(vector1000, criticalValue)
}
